sui_types/src/readable_serde.rs

```diff
@@ -0,0 +1,167 @@
+    data: &[u8],
+    encoding: Encoding,
+) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    if serializer.is_human_readable() {
+        encoding.encode(data).serialize(serializer)
+    } else {
+        Bytes::serialize_as(&data, serializer)
+    }
+}
+
+impl<'de> BytesOrBase64<'de> for ed25519_dalek::Signature {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+        Self: AsRef<[u8]>,
+    {
+        if serializer.is_human_readable() {
+            Encoding::Base64.encode(self.as_ref()).serialize(serializer)
+        } else {
+            <Self as Serialize>::serialize(self, serializer)
+        }
+    }
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        if deserializer.is_human_readable() {
+            let s = String::deserialize(deserializer)?;
+            let value = Encoding::Base64
+                .decode(s)
+                .map_err(to_custom_error::<'de, D, _>)?;
+            Self::try_from(value.as_slice()).map_err(to_custom_error::<'de, D, _>)
+        } else {
+            <Self as Deserialize>::deserialize(deserializer)
+        }
+    }
+}
+
+fn to_custom_error<'de, D, E>(e: E) -> D::Error
+where
+    E: Display,
+    D: Deserializer<'de>,
+{
+    D::Error::custom(format!("byte deserialization failed, cause by: {}", e))
+}
+
+macro_rules! byte_or_hex {
+    ($($len:expr,)+) => {
+        $(
+            impl<'de> BytesOrHex<'de> for [u8; $len]{
+                fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+                where
+                    D: Deserializer<'de>,
+                {
+                    let value = deserialize_from_bytes_or_decode(deserializer, Encoding::Hex)?;
+                    let mut array = [0u8; $len];
+                    array.copy_from_slice(&value[..$len]);
+                    Ok(array)
+                }
+            }
+        )+
+    }
+}
+
+macro_rules! byte_or_base64 {
+    ($($len:expr,)+) => {
+        $(
+            impl<'de> BytesOrBase64<'de> for [u8; $len]{
+                fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+                where
+                    D: Deserializer<'de>,
+                {
+                    let value = deserialize_from_bytes_or_decode(deserializer, Encoding::Base64)?;
+                    let mut array = [0u8; $len];
+                    array.copy_from_slice(&value[..$len]);
+                    Ok(array)
+                }
+            }
+        )+
+    }
+}
+
+enum Encoding {
+    Base64,
+    Hex,
+}
+
+impl Encoding {
+    fn decode(&self, s: String) -> Result<Vec<u8>, anyhow::Error> {
+        Ok(match self {
+            Encoding::Base64 => base64::decode(s)?,
+            Encoding::Hex => hex::decode(s)?,
+        })
+    }
+
+    fn encode<T: AsRef<[u8]>>(&self, data: T) -> String {
+        match self {
+            Encoding::Base64 => base64::encode(data),
```

Comment:
There are issues with the base64 crate in cryptographic contexts (I don't think you're serializing private keys, but you're serializing public ones at least). Please make sure you contact @kchalkias for advice.
[base64ct](https://docs.rs/base64ct/latest/base64ct/) does not have those issues, you may want to use it instead.